### ISSUE TITLE
Ant task keeps report files open on failure

### ISSUE THREAD

**Issue by oehme** [jacoco/jacoco#1160]
### Steps to reproduce

* JaCoCo version: 0.8.6
* Tool integration: Ant (and Gradle, because it uses the Ant task)
* Steps: Run the report generator task on faulty data (e.g. duplicate classes)

### Expected behaviour
The report files do not remain open.

### Actual behaviour

The report files remain open.

This is because the CSV and HTML visitors are [never closed](https://github.com/jacoco/jacoco/blob/66b5fa9216577bfe86fb29e83b031a3cde10cc08/org.jacoco.ant/src/org/jacoco/ant/ReportTask.java#L502) in case of an exception, but only [upon successful completion](https://github.com/jacoco/jacoco/blob/66b5fa9216577bfe86fb29e83b031a3cde10cc08/org.jacoco.ant/src/org/jacoco/ant/ReportTask.java#L497).

**Comment by marchof** [comment]
Thanks @oehme for reporting this! Can you elaborate a bit about a scenario where this results in a actual problem? I would assume that the build aborts in this case anyways. I'm thinking about a test case...

BTW, looks like Maven plugin has the [same issue](https://github.com/jacoco/jacoco/blob/66b5fa9216577bfe86fb29e83b031a3cde10cc08/jacoco-maven-plugin/src/org/jacoco/maven/AbstractReportMojo.java#L182)

**Comment by oehme** [comment]
The problem is that Gradle calls this in-process and the Gradle daemon stays alive across builds. So the files remain locked until the daemon is killed (which users should never have to do).

**Comment by marchof** [comment]
This applies for all resources we open: exec files, class/jar and files written by the reports formatters. Not sure whether there is a robust solution without ty-with-resources (remember, we're on Java 5).

**Comment by oehme** [comment]
There are also effort to create a Maven daemon, so the same problem will apply there.

> Not sure whether there is a robust solution without ty-with-resources

try-with-resources is just shorthand for try-finally. Spotbugs will tell you when you don't close resources, though this particular case might be too well hidden even for that.

**Comment by Godin** [comment]
@oehme do you know if there are automated tests in Gradle that check/catch this?

**Comment by marchof** [comment]
@Godin I can give it a try to better close reporting output resources in case of exceptions.

**Comment by Godin** [comment]
@marchof actually locally I already have changes for this 😉 unfortunately without tests and that's why I was wondering

**Comment by marchof** [comment]
@Godin Same here. I'm curious how you will approach this! 😄

**Comment by oehme** [comment]
The simplest way to verify this is to run your tests on Windows and then try to delete all the files afterwards. If they are still open, it'll fail. Alternatively, on Linux you can use `lsof` to detect open file handles and fail the build if you find any.

**Comment by Godin** [comment]
@oehme before asking I tried

```
	<target name="testReportXmlCloseFile">
		<au:expectfailure expectedMessage="Group name must be supplied">
			<jacoco:report>
				<xml destfile="${temp.dir}/report.xml"/>
			</jacoco:report>
		</au:expectfailure>
		<delete file="${temp.dir}/report.xml"/>
	</target>
```

and indeed can clearly see with breakpoint after `jacoco:report` and prior to `delete`
using `lsof | grep "report.xml"` on macOS.

Was expecting that this will fail on `delete` on Windows in AppVeyor, but it didn't (https://ci.appveyor.com/project/JaCoCo/jacoco/builds/39034510)
I guess because delete was requested from the same process.

Thanks for your response anyway.

**Comment by oehme** [comment]
Yes, the delete would have to be from a different process. I think using lsof is the better alternative.
